extensions-contrib/druid-deltalake-extensions/src/main/java/org/apache/druid/delta/input/DeltaInputSourceReader.java

```diff
@@ -0,0 +1,138 @@
+
+import java.io.IOException;
+import java.util.NoSuchElementException;
+
+/**
+ * A reader for the Delta Lake input source. It initializes an iterator {@link DeltaInputSourceIterator}
+ * for a subset of Delta records given by {@link FilteredColumnarBatch} and schema {@link InputRowSchema}.
+ *
+ */
+public class DeltaInputSourceReader implements InputSourceReader
+{
+  private final io.delta.kernel.utils.CloseableIterator<FilteredColumnarBatch> filteredColumnarBatchCloseableIterator;
+  private final InputRowSchema inputRowSchema;
+
+  public DeltaInputSourceReader(
+      io.delta.kernel.utils.CloseableIterator<FilteredColumnarBatch> filteredColumnarBatchCloseableIterator,
+      InputRowSchema inputRowSchema
+  )
+  {
+    this.filteredColumnarBatchCloseableIterator = filteredColumnarBatchCloseableIterator;
+    this.inputRowSchema = inputRowSchema;
+  }
+
+  @Override
+  public CloseableIterator<InputRow> read()
+  {
+    return new DeltaInputSourceIterator(filteredColumnarBatchCloseableIterator, inputRowSchema);
+  }
+
+  @Override
+  public CloseableIterator<InputRow> read(InputStats inputStats)
+  {
+    return new DeltaInputSourceIterator(filteredColumnarBatchCloseableIterator, inputRowSchema);
+  }
+
+  @Override
+  public CloseableIterator<InputRowListPlusRawValues> sample()
+  {
+
+    CloseableIterator<InputRow> inner = read();
+    return new CloseableIterator<InputRowListPlusRawValues>()
+    {
+      @Override
+      public void close() throws IOException
+      {
+        inner.close();
+      }
+
+      @Override
+      public boolean hasNext()
+      {
+        return inner.hasNext();
+      }
+
+      @Override
+      public InputRowListPlusRawValues next()
+      {
+        DeltaInputRow deltaInputRow = (DeltaInputRow) inner.next();
+        return InputRowListPlusRawValues.of(deltaInputRow, deltaInputRow.getRawRowAsMap());
+      }
+    };
+  }
+
+  private static class DeltaInputSourceIterator implements CloseableIterator<InputRow>
+  {
+    private final io.delta.kernel.utils.CloseableIterator<FilteredColumnarBatch> filteredColumnarBatchCloseableIterator;
+
+    private io.delta.kernel.utils.CloseableIterator<Row> currentBatch = null;
+    private final InputRowSchema inputRowSchema;
+
+    public DeltaInputSourceIterator(
+        io.delta.kernel.utils.CloseableIterator<FilteredColumnarBatch> filteredColumnarBatchCloseableIterator,
+        InputRowSchema inputRowSchema
+    )
+    {
+      this.filteredColumnarBatchCloseableIterator = filteredColumnarBatchCloseableIterator;
+      this.inputRowSchema = inputRowSchema;
+    }
+
+    @Override
+    public boolean hasNext()
+    {
+      while (currentBatch == null || !currentBatch.hasNext()) {
+        if (!filteredColumnarBatchCloseableIterator.hasNext()) {
+          return false; // No more batches or records to read!
+        }
+        currentBatch = filteredColumnarBatchCloseableIterator.next().getRows();
```

Comment:
one qn: do we want to check if the `currentBatch` can return `hasNext` to true? because it could contain zero rows due to the selection vector.